Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,381 @@
+    // FIXME: Is a select element whose multiple content attribute is absent.
+    
+    return false;
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#non-searchable-subtree
+static bool isNonSearchableSubtree(const Node& node)
+{
+    const Node* traversingNode = &node;
+    while (traversingNode) {
+        if (isSearchInvisible(*traversingNode))
+            return true;
+        traversingNode = traversingNode->parentOrShadowHostNode();
+    }
+    return false;
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#nearest-block-ancestor
+static const Node& nearestBlockAncestor(const Node& node)
+{
+    const Node* currentNode = &node;
+    while (currentNode) {
+        if (!currentNode->isTextNode() && currentNode->renderer() && currentNode->renderer()->style().isDisplayBlockLevel())
+            return *currentNode;
+        currentNode = currentNode->parentInComposedTree();
+    }
+    return node.document();
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#get-boundary-point-at-index
+static std::optional<BoundaryPoint> boundaryPointAtIndexInNodes(unsigned index, const Vector<RefPtr<Node>>& textNodeList, BoundaryPointIsAtEnd isEnd)
+{
+    unsigned currentLength = 0;
+    
+    for (auto currentNode : textNodeList) {
+        unsigned nodeEnd = currentLength + currentNode->length();
+        if (isEnd == BoundaryPointIsAtEnd::Yes)
+            nodeEnd++;
+        if (nodeEnd > index)
+            return BoundaryPoint(*currentNode, index - currentLength);
+        currentLength += currentNode->length();
+    }
+    return std::nullopt;
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#is-at-a-word-boundary
+static bool indexIsWordBoundary(const String& string, unsigned index)
 {
-    return &range.start.container->rootNode() == &range.end.container->rootNode()
-        ? WTFMove(range) : SimpleRange { range.start, range.start };
+    UBreakIterator* it = wordBreakIterator(string);
+    return ubrk_isBoundary(it, index);
 }
 
-static std::optional<SimpleRange> rangeOfString(const String& target, Document& document)
+// https://wicg.github.io/scroll-to-text-fragment/#visible-text-node
+static bool isVisibleTextNode(const Node& node)
 {
-    if (target.isEmpty())
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-node-list
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, const SimpleRange& searchRange, Vector<RefPtr<Node>>& nodes, WordBounded wordStartBounded, WordBounded wordEndBounded)
+{
+    String searchBuffer;
+    std::optional<BoundaryPoint> start;
+    std::optional<BoundaryPoint> end;
+    
+    if (nodes.size() < 1)
         return std::nullopt;
+    
+    StringBuilder searchBufferBuilder;
+    for (auto node : nodes)
+        searchBufferBuilder.append(downcast<Text>(*node).data());
+    searchBuffer = searchBufferBuilder.toString();
```

Comment:
This is going to be one massive string...

Reply:
Can we use `SearchBuffer` in `TextIterator.h` instead?
Maybe add a FIXME for now.